ENT-6286: OS: Upgrade version of Artemis to 2.19.1

vkolomeyko opened the pull request:
Fixes DDoS attack mentioned on the Jira ticket.

PR upgrades Artemis library to version 2.19.1.
This is our own release of Apache Artemis library which has vulnerability fix for v2.20 applied.

**_Breaking changes discovered during Artemis upgrade:_**
1. When the queue is created as temporary, it needs to explicitly be specified as non-durable.
2. By default, Artemis Client performs Host DNS name check against the certificate presented by the server. Our TLS certificates fail this check and this verification has to be explicitly disabled, see use of: `TransportConstants.VERIFY_HOST_PROP_NAME`.
3. Artemis Server now caches login attempts, even unsuccessful ones. When we add RPC users dynamically via DB insert this may have an unexpected outcome if the user with the same `userName` and `password` was not available previously.
To workaround permissions changing dynamically, authorization and authentication caches had to be disabled.
4. When computing `maxMessageSize`, the size of the headers content is now taken into account as well.
5. Artemis handling of start-up errors has changed. E.g. when the port is already bound.
6. A number of deprecated APIs like: `createTemporaryQueue`, `failoverOnInitialAttempt`, `NullOutputStream`, `CoreQueueConfiguration`.
7. Log warning message is produced like: `AMQ212080: Using legacy SSL store provider value: JKS. Please use either 'keyStoreType' or 'trustStoreType' instead as appropriate.`
8. As reported by QA, Artemis now produces more audit logging more details [here](https://r3-cev.atlassian.net/browse/ENT-6540). Log configuration been adjusted to reduce such output.

Commit:
ENT-6286: Upgrade version of Artemis to `2.19.0`

Note: It **does not** fix DDoS attacked mention on the Jira ticket, but rather an incremental step of Artemis upgrade which is before the change was mora than 3 years behind latest version.

Commit:
ENT-6286: Use `java.security.cert.X509Certificate`

Commit:
ENT-6286: Fix compilation errors and deprecated API use

Commit:
ENT-6286: Making Detekt happier

Commit:
ENT-6286: Change Artemis connector(i.e. client) configuration

So that it will not be performing verification of the host from the server provided certificate.

Commit:
ENT-6286: Remove deprecated API use

Commit:
ENT-6286: Remove no longer user parameter

More details on: https://issues.apache.org/jira/browse/ARTEMIS-2523

Commit:
ENT-6286: Remove deprecated API use

Commit:
ENT-6286: Explicitly make temporary queue non-durable

Commit:
ENT-6286: Remove deprecated API use

Commit:
Merge remote-tracking branch 'origin/release/os/4.8' into vkolomeyko/ENT-6286-artemis-upgrade

Commit:
ENT-6286: Fix `UserValidationPluginTest`

Commit:
ENT-6286: Fix `ProtonWrapperTests`

Commit:
ENT-6286: Fix `AuthDBTests`

Apparently Artemis now caches login credentials even for failed logins. And there is no easy way to tweak this. So it is important to only try to logon when credentials are in place.

Commit:
ENT-6286: Better fix for `ProtonWrapperTests`

Commit:
ENT-6286: Fix start-up sequence when port is already bound

Commit:
ENT-6286: Make `P2PMessagingClient` to die on connection failure (configurable)

Commit:
ENT-6286: Improve the way of detecting binding error

Commit:
ENT-6286: Fix for `ArtemisMessagingTest`

Commit:
ENT-6286: Eliminate start-up warnings by using new recommended settings

"AMQ212080: Using legacy SSL store provider value: JKS. Please use either 'keyStoreType' or 'trustStoreType' instead as appropriate."

Commit:
ENT-6286: Using consistent style for `QueueConfiguration` creation

Commit:
ENT-6286: Making Detekt happier

vkolomeyko commented:
@mnesbit @dimosr - since CI checks are green, inviting you to review my work so that you can get an idea about the changes I had to make.
I am going to make similar Artemis upgrade in ENT as well and then test with it to get extra confidence.

dimosr commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/ArtemisMessagingClient.kt:
```diff
@@ -66,7 +66,6 @@ class ArtemisMessagingClient(private val config: MutualSslConfiguration,
                 retryInterval = messagingServerConnectionConfig.retryInterval().toMillis()
                 retryIntervalMultiplier = messagingServerConnectionConfig.retryIntervalMultiplier()
                 maxRetryInterval = messagingServerConnectionConfig.maxRetryInterval(isHA).toMillis()
-                isFailoverOnInitialConnection = messagingServerConnectionConfig.failoverOnInitialAttempt(isHA)
```
is there any replacement API we should be using here? Or does failover happen always now?

After a quick look at the `ServerLocator`, I couldn't really tell.

vkolomeyko replied:
They have deprecated `isFailoverOnInitialConnection` as part of: https://issues.apache.org/jira/browse/ARTEMIS-2523
without providing any replacement as this setting no longer makes sense.

Hence, my removal of this setting in C4 code.

dimosr replied:
ah, interesting. I thought this is a toggle for failover between multiple brokers. But, reading what you shared seems like this is done solely based on the addresses provided.

Thanks for the pointer.

dimosr commented on node/src/integration-test/kotlin/net/corda/node/AuthDBTests.kt:
```diff
@@ -158,19 +158,16 @@ class AuthDBTests : NodeBasedTest(cordappPackages = CORDAPPS) {
 
     @Test(timeout=300_000)
-	fun `Add new users dynamically`() {
-        assertFailsWith(
-                ActiveMQSecurityException::class,
-                "Login with incorrect password should fail") {
-            client.start("user2", "bar").close()
-        }
```
I see the initial assertion on failure was removed. Is there a problem with keeping this? Doesn't seem too bad checking it will fail to login before adding the user.

vkolomeyko replied:
Artemis Server now caches login attempts, even unsuccessful ones. When we add RPC users dynamically via DB insert this may have an unexpected outcome if the user with the same userName and password was not available previously.

Also we do not expose relevant caching properties within Artemis using NodeDriver code, hence the easiest way to fix the test was to remove initial unsuccessful logon.

dimosr replied:
Should we maybe disabling this caching (is it even possible) to retain the same UX as before?

(When reading the PR description, I thought that's what we did, but then forgot to the check in the PR)

vkolomeyko replied:
Actually disabling the Artemis Cache may not be such a bad idea. Let me investigate a possibility of doing that.
Then I should be able to restore the test how it originally was.

vkolomeyko replied:
This has been done in the latest commit.
I have restored `AuthDBTests` to how it previously was.

dimosr reviewed (commented):
Looks fine to me, just a couple of minor questions

Commit:
ENT-6286: Effectively disable RPC authentication and authorization cache

As permissions may change dynamically

dimosr reviewed (dismissed):
LGTM

Commit:
ENT-6286: More fixes to dynamic user population logic

If the user is removed mid-session, given that Artemis cache is now disabled, it will be Artemis (and not Shiro) rejecting the user.

To receive response from Artemis we need to switch to blocking non-durable send mode on the RPC Client side. Furthermore, once Artemis rejected the user - the producer associated with the session becomes closed and can no longer be used for anything, e.g. closing the `Observables` so this has to be done in the lenient fashion.

Commit:
Merge remote-tracking branch 'origin/release/os/4.8' into vkolomeyko/ENT-6286-artemis-upgrade

Commit:
ENT-6484: Eliminate warning message introduced

Commit:
Merge remote-tracking branch 'origin/release/os/4.8' into vkolomeyko/ENT-6286-artemis-upgrade

vkolomeyko commented on core-deterministic/testing/data/src/test/kotlin/net/corda/deterministic/data/GenerateData.kt:
```diff
@@ -23,6 +24,25 @@ import kotlin.reflect.jvm.jvmName
 /**
  * Use the JUnit framework to generate a JAR of test data.
  */
+@Ignore("""Need to discuss with Chris Rankin. After Artemis upgrade this test produces:
```
@chrisr3 - we need to discuss this when you are back.
This is related to https://issues.apache.org/jira/projects/ARTEMIS/issues/ARTEMIS-3420

Commit:
Merge remote-tracking branch 'origin/release/os/4.9' into vkolomeyko/ENT-6286-artemis-upgrade

Commit:
ENT-6388: Update `Crash` shell to version 1.7.5

This is a newly built internal version of the library which in particular upgrades `org.apache.mina:mina-core` to version `2.0.22` eliminating NexusIQ vulnerability.

Commit:
Merge remote-tracking branch 'origin/vkolomeyko/ENT-6388-crash-upgrade' into vkolomeyko/ENT-6286-artemis-upgrade

Commit:
ENT-6540: Remove extra logging produced by `org.apache.activemq.audit`

Commit:
Merge remote-tracking branch 'origin/release/os/4.9' into vkolomeyko/ENT-6286-artemis-upgrade

Commit:
ENT-6286: Update dependency of Artemis to `2.19.1`

This is our very own version of Apache Artemis built in-house.

Commit:
Merge remote-tracking branch 'origin/release/os/4.9' into vkolomeyko/ENT-6286-artemis-upgrade

# Conflicts:
#	build.gradle

mnesbit reviewed (approved):
OK
